go/codegen/experimental/codegen_step.go

```diff
@@ -24,37 +28,130 @@ import (
 	pb "sqlflow.org/sqlflow/go/proto"
 )
 
-// TODO(sneaxiy): implement this method to distinguish whether
-// a model is a XGBoost model.
-func isTrainedXBoostModel(modelName string) bool {
-	return true
-}
-
-func generateStepCode(sqlStmt ir.SQLFlowStmt, stepIndex int, session *pb.Session) (string, error) {
+func generateStepCodeAndImage(sqlStmt ir.SQLFlowStmt, stepIndex int, session *pb.Session, sqlStmts []ir.SQLFlowStmt) (string, string, error) {
 	switch stmt := sqlStmt.(type) {
 	case *ir.TrainStmt:
-		return generateTrainCode(stmt, stepIndex, session)
+		return generateTrainCodeAndImage(stmt, stepIndex, session)
 	case *ir.PredictStmt:
-		return generatePredictCode(stmt, stepIndex, session)
+		return generatePredictCodeAndImage(stmt, stepIndex, session, sqlStmts)
 	case *ir.NormalStmt:
-		return GenerateNormalStmtStep(string(*stmt), session, stepIndex)
+		code, err := generateNormalStmtStep(string(*stmt), stepIndex, session)
+		return code, "", err
 	default:
-		return "", fmt.Errorf("not implemented stmt execution type %v", stmt)
+		return "", "", fmt.Errorf("not implemented stmt execution type %v", stmt)
+	}
+}
+
+func generateTrainCodeAndImage(trainStmt *ir.TrainStmt, stepIndex int, session *pb.Session) (string, string, error) {
+	isXGBoost := isXGBoostEstimator(trainStmt.Estimator)
+	if isXGBoost {
+		code, err := XGBoostGenerateTrain(trainStmt, stepIndex, session)
+		if err != nil {
+			return "", "", err
+		}
+		return code, trainStmt.ModelImage, nil
+	}
+	return "", "", fmt.Errorf("not implemented estimator type %s", trainStmt.Estimator)
+}
+
+func generatePredictCodeAndImage(predStmt *ir.PredictStmt, stepIndex int, session *pb.Session, sqlStmts []ir.SQLFlowStmt) (string, string, error) {
+	trainStmt := findModelGenerationTrainStmt(predStmt.Using, stepIndex, sqlStmts)
+	image := ""
+	isXGBoost := false
+	if trainStmt != nil {
+		image = trainStmt.ModelImage
+		isXGBoost = isXGBoostEstimator(trainStmt.Estimator)
+	} else {
+		meta, err := getModelMetadata(session, predStmt.Using)
+		if err != nil {
+			return "", "", err
+		}
+		image = meta.imageName()
```

Comment:
What if meta is not loaded?

Reply:
If meta is not loaded, the error would be returned.